src/MainWindow.vala

```diff
@@ -1077,14 +1079,7 @@ namespace Terminal {
                 }
             });
 
-            /* This signal is not emitted when the .bashrc is missing or does not force terminal to
-             * set its title. So we cannot detect path changes in the shell. Tab name remains "Terminal".
-             */
-            t.window_title_changed.connect (() => {
-                if (t == current_terminal) {
-                    title = t.window_title;
-                }
-
+            t.contents_changed.connect (() => {
                 schedule_name_check ();
```

Comment:
This gets called "whenever the visible appearance of the terminal has changed". This means it is called on every keystroke and also when a program which outputs a log messages or graphical program like htop changes the contents of the terminal. So this means it is called approximately 50-500 times per minute for normal terminal usage. So it should be made sure that this calls are not too expensive.

At the moment this callback eventually calls `save_opened_terminals`, which is relatively expensive operation. For my machine the CPU usage more than doubled when using vim.

On counter measure would be to check if the terminal has a running foreground process and return immediately if this is the case.

Reply:
Maybe this should also only be run on user input and not if the contents of the terminal are changed by a program.

Reply:
Denouncing the `contents_changed` by ~500ms could probably also help.

Reply:
Another idea: Doesn't changing the directory always require to press <kbd>Enter</kdb>? Wouldn't be the simplest solution to schedule the name_check every time <kbd>Enter</kdb> is pressed and no foreground process is running?

Reply:
The cwd can be changed by scripts, which are not foreground processes. If there is a way of detecting when a bash script is running (or detecting when a script inserts `Enter` then that would be more efficient

Reply:
If the script is run in a sub-process (which is in my understanding equivalent to foreground process) like this
```
bash script.sh or ./script.sh
```
it shouldn't be able to change the CWD. But if somebody sources a script which changes the cwd script like this

```
source script.sh
```
yes then it could the change the CWD. This is probably really rare so you have to decide if it is worth taking into acount..

Also there is already some kind of precedent set with the `remember_position` which relies on the <kbd>Enter</kbd> key being pressed. Using the `contents_changed` signal introduces a second mechanism for similar issue: Both want to check if a command was executed. IMO, ideally both should rely on the same mechanism.